src/nvim/channel.c

```diff
@@ -834,3 +836,55 @@ Array channel_all_info(void)
   });
   return ret;
 }
+
+void invoke_prompt_callback(void)
+{
+    typval_T rettv;
+    typval_T argv[2];
+    char_u *text;
+    char_u *prompt;
+    linenr_T lnum = curbuf->b_ml.ml_line_count;
+
+    // Add a new line for the prompt before invoking the callback, so that
+    // text can always be inserted above the last line.
+    ml_append(lnum, (char_u  *)"", 0, false);
+    curwin->w_cursor.lnum = lnum + 1;
+    curwin->w_cursor.col = 0;
+
+    if (curbuf->b_prompt_callback.type == kCallbackNone
```

Comment:
Checking  `curbuf->b_prompt_callback.type == kCallbackNone` should be enough.  There is no point to set `callback.type = kCallbackFuncref` without an actual funcref.

Reply:
@bfredl No, this check is necessary because you can call something like `:call prompt_setcallback(bufnr(''), '')`. If we check only `curbuf->b_prompt_callback.type=kCallbackNone`, `E117: Unknown function: ` occurs.

Reply:
>  However, in the case of :call prompt_setcallback(bufnr(''),''), vim performs the same check so that E117 does not occur. 

Which simply implies that `prompt_setcallback(bufnr(''),'')` should set `kCallbackNone`. The entire point of having callback being an enum is that the enum label should indicate what state the callback actually is in. If both kCallbackNone and kCallbackFuncref can be used used to indicate the valid state of "no callback", both labels become quite meaningless (and sounds like a recipe for future bugs).

> I think that is good if it is a story of changing callback_from_typval() not to allow empty string. But shouldn't it be another PR if we make that change?

This is unrelated to what I suggested. `.funcref=""`, while unfortunate, is never confused for a valid "no callback" in any existing code, what I know of.